ast/compile.go

```diff
@@ -1772,8 +1772,8 @@ func (ci *ComprehensionIndex) String() string {
 
 func buildComprehensionIndices(dbg debug.Debug, arity func(Ref) int, candidates VarSet, rwVars map[Var]Var, node interface{}, result map[*Term]*ComprehensionIndex) uint64 {
 	var n uint64
+	cpy := candidates.Copy()
```

Comment:
I realized that nested bodies will see candidates from subsequent expressions because of how `WalkBodies` and the for-loop below interact... for example:

```rego
x = 7
y = {1 | ["foo" | ...]} # array comprehension nested inside set comprehension
z = 8
```

In this case, the candidate set will contain x and z when the nested array comprehension body is processed.

This shouldn't be a problem in practice because if the nested array comprehension does not close over `z` then it won't matter. If it does close over `z` then `z = 8` will be reordered before the outer set comprehension by an earlier stage.